### ISSUE TITLE
[TODO] 2026

### ISSUE THREAD

**Issue by gongyisheng** [gongyisheng/pretrain#14]
- [x] (not doing) [kernel/triton] flashattn kernel attn_mask support 、
- [x] (not doing) [kernel] fuse rope fansy index to kernel
- [x] [model/components] MLA support
- [ ] [model/components] linear attn support
- [ ] [model/components] XSA support
- [ ] [model/components] sliding window attention support
- [ ] [model/components] attention sink support
- [ ] [mode/components] attn residual support for MoE
- [x] [optm] muon optimizer
- [x] [quant] in8 support 
- [x] [quant] in4 support
- [x] [quant] fp8 support
- [ ] [quant] fp4 support
- [x] [quant] QAT (quantilization aware training) support
- [x] [model/components] more activation support (GeLU, ReLU, GeLU^2, ReLU^2, SwiGLU^2)
- [x] [model/components] MoE implementation
- [x] [kernel/triton] MoE performance
- [x] [experiment] dropout
- [x] [experiment] ffn/attn only dropout
- [x] [experiment] embd only dropout
- [x] [experiment] tie word embeddings
- [x] [experiment] attn comparison (MHA, GQA, MLA, etc)
- [x] [experiment] muon vs adamw
- [x] [experiment] MoE training, expert num
- [x] [experiment] MoE training, aux_loss_coefficient
- [x] [experiment] MoE training, shared expert
- [x] [experiment] activation comparison
- [x] [experiment] MoE activation comparison
- [x] [experiment] quant comparison
- [ ] [experiment] ffn-only quant comparison
- [ ] [experiment] MoE ffn layer quant comparison
- [ ] [experiment] RoPE theta
- [ ] [experiment] attn residual experiment on dense / moe again
- [ ] [debug] per-layer gradient
- [ ] [debug] pre-layer residual similarity
- [ ] [debug] attn score visualization
- [x] [debug] grad norm spike debug - bad data batch
- [x] [config] refactor config, unused feature config default to None
- [x] [config] separate config validation

**Comment by gongyisheng** [comment]
- [x] [experiment] gorkking https://arxiv.org/abs/2201.02177
- [x] [experiment] superbpe https://arxiv.org/pdf/2503.13423

**Comment by gongyisheng** [comment]
- [ ] [experiment] muonH  
- [ ] [experiment] shampoo
- [ ] [experiment] fixed rms weight https://arxiv.org/pdf/2606.16899
- [ ] [experiment] stable latent moe (add rms norm)
- [ ] [experiment] pre-head norm muon, split gate-up muon
- [ ] [experiment] (kimi-k3) quantile balancing
- [ ] [experiment] (kimi-k3) SiTU-GLU
- [ ] [experiment] (kimi-k3) kda, gated mla
- [ ] [experiment] LR annealing
- [ ] [quant] stochastic rounding
- [ ] [quant] Hadamard / SmoothQuant (Outlier handling)
- [ ] [quant] depth sensitivity
